### DISCUSSION TITLE
Variable visibility problem in inner functions

### DISCUSSION THREAD

**Discussion by gery** [vlang/v#11263] [General]
I'm developing a websocket game in vlang, I face this problem:

    mut data := 'testMessageData'

	s.on_message_ref(fn (mut ws websocket.Client, msg &websocket.Message, mut m websocket.Server) ? {
		for i, _ in m.clients {
			mut c := m.clients[i]
                       if msg.payload == 'get-map'.bytes() {
                           c.client.write(data.bytes(), websocket.OPCode.text_frame) or { panic(err) }
                       }
		}
	}, s)

The data variable is not visible inside. There are no global variables in Vlang. Is there any solution or workaround for this problem?

compile error message:

./GameServer.v:97:32: error: undefined ident: `data`
   95 |             if msg.payload == 'get-map'.bytes() {
   96 |                 //create_game().get_map_string()
   97 |                 c.client.write(data.bytes(), websocket.OPCode.text_frame) or { panic(err) }
      |                                ~~~~
   98 |             }
   99 |         }

Anyway, I found Vlang very COOL ;)

**Comment by JalonSolov** [comment]
You have an anonymous function.  You need a closure or a regular function and pass the value as a parameter.

https://github.com/vlang/v/blob/master/doc/docs.md#anonymous--higher-order-functions

**Comment by gery** [comment]
Thanks for the fast answer.
In a regular function, i cannot add extra parameter,
with a closure i get this compile time error:

gery@gery-XPS-13-9360:~/projects/vlang/game$ v ./GameServer.v 

/tmp/v/GameServer.18085048805967199074.tmp.c:6290: error: pointer expected

(Use `v -cg` to print the entire error message)

builder error: 

C error. This should never happen.

If you were not working with C interop, this is a compiler bug, please report the bug using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

My full source code (96 lines) is here:

```
module main

import vweb
import os
import rand
import net.websocket
import term

const (
	map_size  = 10 
)

struct App {
    vweb.Context   
}

[heap]
struct Game{
  mut:
   name string
   map [][]int
}

fn (mut g Game) generate_map() {
    g.map  =  [][]int{len: map_size, init: []int{len: map_size}}
    for y in 0 .. map_size-1 {
       for x in 0 .. map_size-1 {
          g.map[x][y]=rand.intn(2) 
          print(g.map[x][y])
       }
    }
}

fn (g Game) get_map_string() string {
    mut map_string := ''
    for y in 0 .. map_size-1 {
       for x in 0 .. map_size-1 {
          map_string += g.map[x][y].str()
       }
       map_string += '|'
    }
    return map_string
}

fn main() {
    mut game := Game{
		name: 'first'
	}
    game.generate_map()
    go vweb.run(&App{}, 8080)
    println('starting websocket')
    start_websocket_server(mut game) ?
}

pub fn (mut app App) index() vweb.Result {
    filename := './index.html'
    data := os.read_file(filename) or {
        panic('error reading file $filename')
    }
	return app.html(data)
}


fn start_websocket_server( mut game Game) ? {
	mut s := websocket.new_server(.ip6, 30000, '')
    println('game map:'+game.get_map_string())

	// Make that in execution test time give time to execute at least one time
	s.ping_interval = 100
	s.on_connect(fn (mut s websocket.ServerClient) ?bool {
		// Here you can look att the client info and accept or not accept
		// just returning a true/false
		if s.resource_name != '/' {
			return false
		}
		return true
	}) ?
    mut ref := &game
	// on_message_ref, broadcast all incoming messages to all clients except the one sent it
	s.on_message_ref(fn [ref](mut ws websocket.Client, msg &websocket.Message, mut m websocket.Server) ? {
		for i, _ in m.clients {
			mut c := m.clients[i]
            if msg.payload == 'get-map'.bytes() {
                c.client.write( (*ref).get_map_string().bytes(), websocket.OPCode.text_frame) or { panic(err) }
            }
		}
	}, s)

	s.on_close(fn (mut ws websocket.Client, code int, reason string) ? {
		println(term.green('client ($ws.id) closed connection'))
	})
	s.listen() or { println(term.red('error on server listen: $err')) }
	unsafe {
		s.free()
	}
}

**Comment by gery** [comment]
I tried a simpler example:
```
struct Game{
  mut:
   name string
}

mut game := Game{
    name: 'first'
}

print_counter2 := fn [game]( name2 string) {
	println(game.name)
	println(name2)
}

print_counter2('a') 
game.name='second'
print_counter2('b') 
```
The result is:

v ./test.v 
./test 

/tmp/v/test.2488586026076545758.tmp.c:1862: at anon_fn_db322b848a87d5fe_string_142: RUNTIME ERROR: invalid memory access
/tmp/v/test.2488586026076545758.tmp.c:10398: by main__main
/tmp/v/test.2488586026076545758.tmp.c:10756: by main


OS: ubuntu linux,  V version:  0.2.2 c9d6242
